File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter13/DialSketch/DialSketch/Dial.cpp
#include "pch.h"
#include "Dial.h"

using namespace DialSketch;

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Input;

Dial::Dial()
{
    this->ManipulationMode = ManipulationModes::Rotate;
}

void Dial::OnManipulationStarting(ManipulationStartingRoutedEventArgs^ args)
{
    args->Pivot = ref new ManipulationPivot(Point((float)(this->ActualWidth / 2), 
                                                  (float)(this->ActualHeight / 2)), 
                                            48);
}

void Dial::OnManipulationDelta(ManipulationDeltaRoutedEventArgs^ args)
{
    this->Value = max(this->Minimum,
                  min(this->Maximum, this->Value + args->Delta.Rotation));
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter13/DialSketch/DialSketch/Dial.cs
using System;
using Windows.Foundation;
using Windows.UI.Xaml.Controls.Primitives;
using Windows.UI.Xaml.Input;

namespace DialSketch
{
    public class Dial : RangeBase
    {
        public Dial()
        {
            ManipulationMode = ManipulationModes.Rotate;
        }

        protected override void OnManipulationStarting(ManipulationStartingRoutedEventArgs args)
        {
            args.Pivot = new ManipulationPivot(new Point(this.ActualWidth / 2, 
                                                         this.ActualHeight / 2), 
                                               48);
            base.OnManipulationStarting(args);
        }

        protected override void OnManipulationDelta(ManipulationDeltaRoutedEventArgs args)
        {
            this.Value = Math.Max(this.Minimum, 
                         Math.Min(this.Maximum, this.Value + args.Delta.Rotation));

            base.OnManipulationDelta(args);
        }
    }
}
